File: frontend/src/api/review.ts
export type QuestionType = 'context_fill' | 'meaning_choice'

export interface ReviewQuestion {
  id: string
  type: QuestionType
  prompt: string
  options: string[]
  lemma: string
  phonetic: string | null
  meanings: string[]
  context: string
  source_book_title: string | null
  source_chapter_title: string | null
}

export interface ReviewSession {
  questions: ReviewQuestion[]
  total_available: number
  due_count: number
  scheduled_count: number
  next_review_at: string | null
}

export interface ReviewResult {
  is_correct: boolean
  correct_answer: string
  explanation: string
  wrong_count: number
  review_stage: number
  next_review_at: string
  answered_at: string
}

export interface ReviewStats {
  total_attempts: number
  correct_attempts: number
  accuracy: number
  words_practiced: number
  due_count: number
  scheduled_count: number
  next_review_at: string | null
}

export interface GateReviewCompletion {
  completed_count: number
  repaired_count: number
  next_review_at: string
}

export async function fetchReviewSession(limit = 10): Promise<ReviewSession> {
  return requestJson(`/api/v1/review/session?limit=${limit}`)
}

export async function fetchReviewStats(): Promise<ReviewStats> {
  return requestJson('/api/v1/review/stats')
}

export async function submitReviewAnswer(
  question: ReviewQuestion,
  answer: string,
): Promise<ReviewResult> {
  return requestJson('/api/v1/review/answer', {
    method: 'POST',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify({
      question_id: question.id,
      question_type: question.type,
      prompt: question.prompt,
      answer,
    }),
  })
}

export async function completeGateReview(
  questions: ReviewQuestion[],
  mistakeCounts: Record<string, number>,
): Promise<GateReviewCompletion> {
  return requestJson('/api/v1/review/complete', {
    method: 'POST',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify({
      items: questions.map((question) => ({
        question_id: question.id,
        mistake_count: mistakeCounts[question.id] || 0,
      })),
    }),
  })
}

async function requestJson<T>(url: string, init?: RequestInit): Promise<T> {
  const response = await fetch(url, init)
  if (!response.ok) {
    let message = '训练请求失败，请稍后重试'
    try {
      const body = (await response.json()) as { error?: { message?: string } }
      message = body.error?.message || message
    } catch {
      // Keep the generic fallback when the service returns no structured error.
    }
    throw new Error(message)
  }
  return response.json() as Promise<T>
}

File: backend/app/schemas/review.py
from datetime import datetime
from typing import Literal

from pydantic import BaseModel, Field

QuestionType = Literal["context_fill", "meaning_choice"]


class ReviewQuestion(BaseModel):
    id: str
    type: QuestionType
    prompt: str
    options: list[str]
    lemma: str
    phonetic: str | None
    meanings: list[str]
    context: str
    source_book_title: str | None
    source_chapter_title: str | None


class ReviewSessionResponse(BaseModel):
    questions: list[ReviewQuestion]
    total_available: int
    due_count: int
    scheduled_count: int
    next_review_at: datetime | None


class SubmitReviewRequest(BaseModel):
    question_id: str
    question_type: QuestionType
    prompt: str = Field(min_length=1, max_length=5000)
    answer: str = Field(min_length=1, max_length=128)


class SubmitReviewResponse(BaseModel):
    is_correct: bool
    correct_answer: str
    explanation: str
    wrong_count: int
    review_stage: int
    next_review_at: datetime
    answered_at: datetime


class ReviewStatsResponse(BaseModel):
    total_attempts: int
    correct_attempts: int
    accuracy: float
    words_practiced: int
    due_count: int
    scheduled_count: int
    next_review_at: datetime | None


class CompleteGateItem(BaseModel):
    question_id: str
    mistake_count: int = Field(ge=0, le=100)


class CompleteGateReviewRequest(BaseModel):
    items: list[CompleteGateItem] = Field(min_length=1, max_length=30)


class CompleteGateReviewResponse(BaseModel):
    completed_count: int
    repaired_count: int
    next_review_at: datetime
